File: leetcode/medium/simplify-path.py
def simplifyPath(path: str) -> str:
    stack = []
    
    for dir in list(path.split("/")):
        if dir == "":
            continue
        elif dir == ".":
            continue
        elif dir == "..":
            if stack:
                stack.pop()
        else:
            stack.append(dir)
            
    
    return "/" + "/".join(stack)
    
    
    
print(simplifyPath("/home/"))
print(simplifyPath("/../"))
print(simplifyPath("/home//foo/"))

File: leetcode/medium/simplify-path.go
package main

import (
	"fmt"
	"strings"
)

func simplifyPath(path string) string {
	stack := make([]string, len(path))
	top := 0

	for _, dir := range strings.Split(path, "/") {
		switch dir {
		case "":
			continue
		case ".":
			continue
		case "..":
			if 0 < top {
				top--
			}
		default:
			stack[top] = dir
			top++
		}
	}

	return "/" + strings.Join(stack[:top], "/")
}

func main() {
	fmt.Println(simplifyPath("/home/"))
	fmt.Println(simplifyPath("/../"))
	fmt.Println(simplifyPath("/home//foo/"))
}
